client/router-hooks.ts

```diff
@@ -0,0 +1,38 @@
+import { useLocation } from 'wouter'
+
+/**
+ * A hook that reads a search param with a given name from `window.location`, and allows changing
+ * it. Changing the search param will re-render the component.
+ *
+ * If there are multiple params with the same name (which is apparently supported by the standard),
+ * we return the first one.
+ *
+ * @example
+ *
+ * export const UserList = () => {
+ *   // /users?page=3
+ *   const [page, setPage] = useLocationSearchParam('page')
+ *   console.log(page) // '3'
+ *   return <button onClick={() => setPage('4')}>Go to page 4</button>
+ * }
+ */
+export const useLocationSearchParam = (
+  name: string,
+): [value: string, setValue: (value: string) => void] => {
+  const [location, setLocation] = useLocation()
+  const searchParams = new URLSearchParams(window.location.search)
+  const searchValue = searchParams.get(name) ?? ''
+
+  const setLocationSearch = (value: string) => {
+    if (value) {
```

Comment:
if/when you memoize the above value you'll want to create a copy of the params here before mutating (honestly, you should have done this even when things weren't memoized imo). Think you can do that via:
```ts
const params = new URLSearchParams(Array.from(searchParams.entries()))
```